README
This is the README for CPAN::SQLite, which is used for setting up,
maintaining, and searching through a local CPAN database consisting
of information in the three main CPAN index files:
  $CPAN/modules/03modlist.data.gz
  $CPAN/modules/02packages.details.txt.gz
  $CPAN/authors/01mailrc.txt.gz
DBD::SQLite is used as the database engine.

Installation proceeds through the usual
   perl Makefile.PL
   $MAKE
   $MAKE test
   $MAKE install
or, if Module::Build is available,
   perl Build.PL
   perl Build
   perl Build test
   perl Build install

The package consists of a number of modules:
   CPAN::SQLite: the top-level package
   CPAN::SQLite::Index: set up and maintain the database
   CPAN::SQLite::Info: extract information from the CPAN indices
   CPAN::SQLite::META: helper module for CPAN.pm integration
   CPAN::SQLite::State: gather information on the state of the database
   CPAN::SQLite::Search: methods to search the database
   CPAN::SQLite::Populate: populate the database tables
   CPAN::SQLite::DBI: DBI information for the database
   CPAN::SQLite::DBI::Index: DBI information for indexing the database
   CPAN::SQLite::DBI:Search: DBI information for searching the database
See the pod documentation of the specific modules for details.

Perhaps the easiest way to set up and maintain the database is through
the supplied cpandb script. If we assume that there is a CPAN.pm
client available and configured, the database can be created via
   cpandb --setup
which will create the database file cpandb.sql under the
"cpan_home" defined in CPAN::Config (this is the same location
as where CPAN.pm's Metadata is found). Subsequent updates to
the database can be done through
   cpandb --update
This scenario assumes that updates to the CPAN index files are
handled by CPAN.pm.

The cpandb script also provides some elementary searches:
   cpandb --module My::Module
   cpandb --dist My-Dist
   cpandb --cpanid ME
which will provide some information on, respectively, the
specified CPAN module, distribution, and CPAN author id.
The query terms are assumed to match exactly, and are
case-insensitive.

As of CPAN.pm version 1.88_65, there is experimental support 
within CPAN.pm for using CPAN::SQLite to obtain
information on packages, modules, and authors. One goal
of this is to reduce the memory footprint of the CPAN.pm
shell, as this information is no longer all preloaded into
memory. This can be enabled through
   perl -MCPAN -e shell
   cpan> o conf use_sqlite 1

This version is considered alpha. Please report bugs
and feature requests to CPAN's request tracker:

  http://rt.cpan.org/NoAuth/Bugs.html?Dist=CPAN-SQLite

or directly to the author. Development takes place at

  http://cpan-search.svn.sourceforge.net/viewvc/cpan-search/CPAN-SQLite/

This software is copyright 2006 by Randy Kobes
<r.kobes@uwinnipeg.ca>. Use and
redistribution are under the same terms as Perl itself.


Changes
Revision history for Perl extension CPAN::SQLite.

0.210 Jul 11, 2015
  - Moved from LWP::Simple to HTTP::Tiny

0.209 Jul 8, 2015
  - Removed some additional tests that don't work well on 5.6.2
  - Better handling of parallel testing

0.208 Jul 5, 2015
  - Got rid of [AssertOS]

0.207 Jul 4, 2015
  - Detect HARNESS_OPTIONS and bail out
  - Back to Dist::Zilla

0.206 Jun 29, 2015
  - No more Module::Build
  - Fixed test failures on Windows
  - Typo fixed
     RT#86588: https://rt.cpan.org/Public/Bug/Display.html?id=86588
     (reported by David Steinbrunner)

0.205 Nov 22, 2014
  - Fixed test failures

0.204 Aug 04, 2014
  - Fixed problem with locked database in DBD::SQLite 1.38_01 and higher
     RT#90635: https://rt.cpan.org/Ticket/Display.html?id=90635
     (reported by Alexandr Ciornii)

0.203 Jun 29, 2013
  - Automatically cleanup log files older than 30 days. Use 
    CPAN_SQLITE_LOG_FILES_CLEANUP environment variable to override.
     RT#76943: https://rt.cpan.org/Ticket/Update.html?id=76943
    (reported by Nigel Horne)
     See also RT#36411
  - Fixed POD
     RT#83818 :https://rt.cpan.org/Ticket/Display.html?id=83818
    (reported by Alceu Rodrigues de Freitas Junior)
  - Added additional checks to make sure cpandb.sql is correctly populated
     RT#64628: https://rt.cpan.org/Ticket/Display.html?id=64628
     RT#71323: https://rt.cpan.org/Ticket/Display.html?id=71323
     RT#76197: https://rt.cpan.org/Ticket/Display.html?id=76197
    (reported by Slaven Rezic, Alexandr Ciornii, Randy Stauner)
  - Tests updated
  - "base" replaced with "parent"
  - Minor tweaks

0.202 Jun 25, 2011
  - Fixed test configuration which caused problems on Cygwin
  - Added CPAN root directory creation in case it's missing

0.201 Jun 16, 2011
  - Fixed problem with 3-argument open with '>&' on 5.6.2
  - Updated repository link

0.200 Jun 11, 2011
  - SQLite REGEXP is used instead of custom function to prevent "Use of
    uninitialized value $x in pattern match" warning in modern Perl versions
     RT#67442: https://rt.cpan.org/Ticket/Display.html?id=67442
    (reported by Mike Doherty)
  - "warn"s are replaced with $CPAN::FrontEnd calls
     RT#59691: https://rt.cpan.org/Ticket/Display.html?id=59691
    (reported by Randy J. Ray)

0.199 Dec 29, 2009
  - require CPAN::DistnameInfo 0.09, which normalizes dist() for CGI.pm
    (reported by Pete Armstrong)

0.198 Sep 19, 2009
  - CPAN::SQLite::DBI::Index had an extra right parenthesis
    (reported by Alexey Borzenkov)
  - CPAN::SQLite::META braces bug in oneliner; see
     http://rt.cpan.org/Ticket/Display.html?id=47353
  - add Compress:Zib dependency; see
     http://rt.cpan.org/Ticket/Display.html?id=45783
  - better error reporting in cases where index files aren't fetched; see
      http://rt.cpan.org/Ticket/Display.html?id=47322

0.197 Feb 16, 2009
  - add LWP::Simple to prerequisites of Build.PL, to help CPAN testers:
        http://www.cpantesters.org/show/CPAN-SQLite.html#CPAN-SQLite-0.196

0.196 June 6, 2008
  - turn off creation of log files during indexing procedures
    via setting $ENV{CPAN_SQLITE_NO_LOG_FILES}, as suggested
    by DAGOLDEN:
        http://rt.cpan.org/Ticket/Display.html?id=36411
        
0.195 Feb 16, 2008
  - fix further bugs where sometimes a regex search for
    distributions and/or modules, followed by a distribution
    search for a specific distribution, didn't properly
    populate the CONTAINSMODS field
  - add CPAN::Shell->reload('index') in the t/05meta.t test
    to ensure the database is created, and test that it has
    non-zero size
  - fix bug where sometimes a regex module search, followed
    by a distribution search for a specific distribution,
    didn't properly set the abstract and version fields
  - fix bug in t/05meta_*.t tests so as to correctly compare modules
    in a distribution.
  - honour $Config::CPAN->{index_expire} to decide when to update
    the database
  - improved error checking for the exisitence of the database,
    including also a check on the size, if it exists, when 
    deciding whether to update or create the database.

0.19 Feb 3, 2008
  - fix bug whereby a query for a module, followed by a query for
    the distribution containing that module, would only list the
    module first queried for CONTAINSMODS
  - fix (hopefully) failing tests on Darwin:
       http://rt.cpan.org/Ticket/Display.html?id=28890
    resulting from CPAN::MyConfig not getting picked up
    from PERL5LIB (thanks to Michael Schwern and David Wheeler)
  - have information messages printed out only if CPAN_SQLITE_DEBUG
    is set, as requested at
       http://rt.cpan.org/Ticket/Display.html?id=32740
    (thanks to ZOFFIX)

0.18 Aug 15, 2007
  - as some older versions of DBD::SQLite don't support
    "drop table if exists $table", use an alternate construction
  - add vcmp function in TestSQL.pm to test equality of versions
  - when testing with CPAN.pm, remove the database file
    before testing

0.15 July 26, 2007
  - add a message of when the database was last updated:
      http://rt.cpan.org/Ticket/Display.html?id=27953
  - don't ignore a previously set PERL5LIB in the tests:
      http://rt.cpan.org/Ticket/Display.html?id=28297

0.1  Feb 17, 2007
  - first non-developer release; no code changes from 0.1_04

0.1_04 Jan 11, 2007
  - improve diagnostics of some of the tests using Test::More
  - check that we loaded the correct CPAN::Config in 05meta.t
    (test suggested by Andreas Koenig)
  - add auto_commit variable to the test's CPAN::Config, so
    as to work with CPAN.pm 1.88_69

0.1_03 Dec 31, 2006
  - in the t/05meta.t test, remove an underscore from the version
    strings, as older versions of CPAN::Version seem to have
    problems comparing versions with underscores.
  - don't try to update database if a $db_name-journal file
    exists, indicating another process is updating the database
  - in a regex search for Bundles within CPAN::SQLite::META,

lib/CPAN/SQLite/Index.pm
# $Id: Index.pm 52 2015-07-12 00:49:18Z stro $

package CPAN::SQLite::Index;
use strict;
use warnings;

our $VERSION = '0.210';

use English qw/-no_match_vars/;

use CPAN::SQLite::Info;
use CPAN::SQLite::State;
use CPAN::SQLite::Populate;
use CPAN::SQLite::DBI qw($tables);
use File::Spec::Functions qw(catfile);
use File::Basename;
use File::Path;
use HTTP::Tiny;

unless ($ENV{CPAN_SQLITE_NO_LOG_FILES}) {
  $ENV{CPAN_SQLITE_DEBUG} = 1;
}

our ($oldout);
my $log_file = 'cpan_sqlite_log.' . time;

# This is usually already defined in real life, but tests need it to be set
$CPAN::FrontEnd ||= "CPAN::Shell";

sub new {
  my ($class, %args) = @_;
  if ($args{setup} and $args{reindex}) {
    die "Reindexing must be done on an exisiting database";
  }

  my $self = {index => undef, state => undef, %args};
  return bless $self, $class;
}

sub index {
    my $self = shift;
    my $setup = $self->{'setup'};

    if ($setup) {
        my $db_name = catfile($self->{'db_dir'}, $self->{db_name});
        if (-f $db_name) {
            $CPAN::FrontEnd->myprint("Removing existing $db_name ... ");
            if (unlink $db_name) {
                $CPAN::FrontEnd->myprint("Done.\n");
            } else {
                $CPAN::FrontEnd->mywarn("Failed: $!\n");
            }
        }
    }

    my $log = catfile($self->{'log_dir'}, $log_file);

    unless ($ENV{'CPAN_SQLITE_NO_LOG_FILES'}) {
        $oldout = error_fh($log);
    }

    my $log_cleanup = $ENV{'CPAN_SQLITE_LOG_FILES_CLEANUP'};
    $log_cleanup = 30 unless defined $log_cleanup;
    if ($log_cleanup and $log_cleanup =~ /^\d+$/) {
      if (opendir(my $DIR, $self->{'log_dir'})) {
        my @files = grep { /cpan_sqlite_log\./ } readdir $DIR;
        closedir $DIR;

        @files = grep { -C $_ > $log_cleanup } map { catfile($self->{'log_dir'}, $_) } @files;

        if (@files) {
          $CPAN::FrontEnd->myprint('Cleaning old log files ... ');
          unlink @files;
          $CPAN::FrontEnd->myprint("Done.\n");
        }
      }
    }

    if ($self->{'update_indices'}) {
        $CPAN::FrontEnd->myprint('Fetching index files ... ');
        if ($self->fetch_cpan_indices()) {
            $CPAN::FrontEnd->myprint("Done.\n");
        } else {
            $CPAN::FrontEnd->mywarn("Failed\n");
            return;
        }
    }

    $CPAN::FrontEnd->myprint('Gathering information from index files ... ');
    if ($self->fetch_info()) {
        $CPAN::FrontEnd->myprint("Done.\n");
    } else {
        $CPAN::FrontEnd->mywarn("Failed\n");
        return;
    }

    unless ($setup) {
        $CPAN::FrontEnd->myprint('Obtaining current state of database ... ');
        if ($self->state()) {
            $CPAN::FrontEnd->myprint("Done.\n");
        } else {
            $CPAN::FrontEnd->mywarn("Failed\n");
            return;
        }
    }

    $CPAN::FrontEnd->myprint('Populating database tables ... ');
    if ($self->populate()) {
        $CPAN::FrontEnd->myprint("Done.\n");
    } else {
        $CPAN::FrontEnd->mywarn("Failed\n");
        return;
    }

    return 1;
}

sub fetch_cpan_indices {
  my $self = shift;

  my $CPAN = $self->{CPAN};
  my $indices = {
    '01mailrc.txt.gz' => 'authors',
    '02packages.details.txt.gz' => 'modules',
    '03modlist.data.gz' => 'modules',
  };

  foreach my $index (keys %$indices) {
    my $file = catfile($CPAN, $indices->{$index}, $index);
    next if (-e $file and -M $file < 1);
    my $dir = dirname($file);
    unless (-d $dir) {
      mkpath($dir, 0, oct(755)) or die "Cannot mkpath $dir: $!";
    }
    my @urllist = @{$self->{urllist}};
    foreach my $cpan(@urllist) {
      my $from = join '/', ($cpan, $indices->{$index}, $index);
      if (my $response = HTTP::Tiny->new->get($from)) {
        if ($response->{'success'}) {
          if (open(my $FILE, '>', $file)) {
            binmode $FILE;
            print $FILE $response->{'content'};
            if (close($FILE)) {
              next;
            }
          }
        }
      }
    }
    unless (-f $file) {
      $CPAN::FrontEnd->mywarn("Cannot retrieve '$file'");
      return;
    }
  }
  return 1;
}

sub fetch_info {
  my $self = shift;
  my %wanted = map {$_ => $self->{$_}} qw(CPAN ignore keep_source_where);
  my $info = CPAN::SQLite::Info->new(%wanted);
  $info->fetch_info() or return;
  my @tables = qw(dists mods auths info);
  my $index;
  foreach my $table(@tables) {
    my $class = __PACKAGE__ . '::' . $table;
    my $this = {info => $info->{$table}};
    $index->{$table} = bless $this, $class;
  }
  $self->{index} = $index;
  return 1;
}

sub state {
  my $self = shift;

  my %wanted = map {$_ => $self->{$_}} qw(db_name index setup reindex db_dir);
  my $state = CPAN::SQLite::State->new(%wanted);
  $state->state() or return;
  $self->{state} = $state;
  return 1;
}

sub populate {
  my $self = shift;
  my %wanted = map {$_ => $self->{$_}} qw(db_name index setup state db_dir);
  my $db = CPAN::SQLite::Populate->new(%wanted);
  $db->populate() or return;
  return 1;
}

sub error_fh {
  my $file = shift;
  open(my $tmp, '>', $file) or die "Cannot open $file: $!";
  close $tmp;
# Should be open(my $oldout, '>&', \*STDOUT); but it fails on 5.6.2
  open(my $oldout, '>&STDOUT');
  open(STDOUT, '>', $file) or die "Cannot tie STDOUT to $file: $!";
  select STDOUT; $| = 1;
  return $oldout;
}

sub DESTROY {
  unless ($ENV{CPAN_SQLITE_NO_LOG_FILES}) {
    close STDOUT;
    open(STDOUT, '>&', $oldout) if $oldout;
  }
  return;
}

1;

=head1 NAME

CPAN::SQLite::Index - set up or update database tables.

=head1 VERSION

version 0.210

=head1 SYNOPSIS

 my $index = CPAN::SQLite::Index->new(setup => 1);
 $index->index();

=head1 DESCRIPTION

This is the main module used to set up or update the
database tables used to store information from the
CPAN and ppm indices. The creation of the object

 my $index = CPAN::SQLite::Index->new(%args);

accepts two possible arguments:

=over 3

=item * setup =E<gt> 1

This (optional) argument specifies that the database is being set up.
Any existing tables will be dropped.

=item * reindex =E<gt> value

This (optional) argument specifies distribution names that
one would like to reindex in an existing database. These may
be specified as either a scalar, for a single distribution,
or as an array reference for a list of distributions.

=back

=head1 DETAILS

Calling

  $index->index();

will start the indexing procedure. Various messages
detailing the progress will written to I<STDOUT>,
which by default will be captured into a file
F<cpan_sqlite_log.dddddddddd>, where the extension
is the C<time> that the method was invoked. Error messages
are not captured, and will appear in I<STDERR>.

The steps of the indexing procedure are as follows.

=over 3

=item * fetch index data

The necessary CPAN index files
F<$CPAN/authors/01mailrc.txt.gz>,
F<$CPAN/modules/02packages.details.txt.gz>, and
F<$CPAN/modules/03modlist.data.gz> will be fetched
from the CPAN mirror specified by the C<$cpan> variable
at the beginning of L<CPAN::SQLite::Index>. If you are
using this option, it is recommended to use the
same CPAN mirror with subsequent updates, to ensure consistency
of the database. As well, the information on the locations
of the CPAN mirrors used for Template-Toolkit and GeoIP
is written.

=item * get index information

Information from the CPAN indices is extracted through
L<CPAN::SQLite::Info>.

=item * get state information

Unless the C<setup> argument within the C<new>
method of L<CPAN::SQLite::Index> is specified,
this will get information on the state of the database
through L<CPAN::SQLite::State>.
A comparison is then made between this information
and that gathered from the CPAN indices, and if there's
a discrepancy in some items, those items are marked
for either insertion, updating, or deletion, as appropriate.

=item * populate the database

At this stage the gathered information is used to populate
the database, through L<CPAN::SQLite::Populate>,
either inserting new items, updating
existing ones, or deleting obsolete items.

=back

=head1 SEE ALSO

L<CPAN::SQLite::Info>, L<CPAN::SQLite::State>,
L<CPAN::SQLite::Populate>,
and L<CPAN::SQLite::Util>.
Development takes place on the CPAN-SQLite project
at L<http://sourceforge.net/projects/cpan-search/>.

=head1 AUTHORS

Randy Kobes (passed away on September 18, 2010)

Serguei Trouchelle E<lt>stro@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2006 by Randy Kobes E<lt>r.kobes@uwinnipeg.caE<gt>.

Copyright 2011 by Serguei Trouchelle E<lt>stro@cpan.orgE<gt>.

Use and redistribution are under the same terms as Perl itself.

=cut
